### PR TITLE
Bug 363595:Configuring Agent with Interactive process is failing in HTTPS

### PR DESCRIPTION
**Pull request by aditya-MSFT**
Bug 363595:Configuring Agent with Interactive process is failing in HTTPS

1. Getting details of WinRm configuration on the machine
2. Making sure we are using correct parameters while launching new PSSession for DTAExecutionHost

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 1]
```diff
@@ -615,6 +616,39 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
 
 }
 
+function CreateNewSession()
+{
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: we should definitely test it on all supported OS

**Comment by ajay-arora** [comment]
please mention testing done

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 2]
```diff
@@ -615,6 +616,39 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
...
+        $port = 5985
+        $transport = HTTP          
+    }
+
+    Write-Verbose -Message("Using Port {0} for creating session" -f $port) -Verbose
+        
+    if( $transport -eq "HTTPS")
+    {
+        $sessionOption = New-PSSessionOption -SkipCACheck -SkipCNCheck
+        $session = New-PSSession -ComputerName . -Port $port -SessionOption $sessionOption -UseSSL
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: you have removed credential usage from this ... any reason for that change ?
if no, then I guess its safer to stick with the old tested code

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 3]
```diff
@@ -607,7 +607,8 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
     $exePath = Join-Path -Path $vsRoot -ChildPath $ExeName
     $exePath = "'" + $exePath + "'"
 
-    $session = New-PSSession -ComputerName . -Credential $MachineCredential
+    $session = CreateNewSession
+
     Invoke-Command -Session $session -ErrorAction Continue -ErrorVariable err -OutVariable out -scriptBlock { schtasks.exe /create /TN:DTAConfig /TR:$args /F /RL:HIGHEST /SD:01/01/2050 /SC:ONCE /ST:00:00 ; schtasks.exe /run /TN:DTAConfig } -ArgumentList $exePath
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: can we place this code in a try block and catch all exceptions ?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 4]
```diff
@@ -606,13 +606,52 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
...
+{
+    $winrmconfigDetails = Winrm e  winrm/config/listener -format:pretty |Out-String
+    $xmldoc = [XML]$winrmconfigDetails    
+    $transports = $xmldoc.GetElementsByTagName("cfg:Transport")
+    $ports = $xmldoc.GetElementsByTagName("cfg:Port")
+   
+    if( $ports -ne $null -and  $transports -ne $null -and $ports.Count -gt 0)
+    {
+        $port = $ports[0].InnerText
+        $transport = $transports[0].InnerText
```
Comment (kmkumaran) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: aren't these details already in environment property that you can use? If a user specifies a specific port and protocol shouldn't that be given preference?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 5]
```diff
@@ -606,13 +606,52 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
...
+		Invoke-Command -Session $session -ErrorAction Continue -ErrorVariable err -OutVariable out -scriptBlock { schtasks.exe /create /TN:DTAConfig /TR:$args /F /RL:HIGHEST /SD:01/01/2050 /SC:ONCE /ST:00:00 ; schtasks.exe /run /TN:DTAConfig } -ArgumentList $exePath
 
-    $session = New-PSSession -ComputerName . -Credential $MachineCredential
-    Invoke-Command -Session $session -ErrorAction Continue -ErrorVariable err -OutVariable out -scriptBlock { schtasks.exe /create /TN:DTAConfig /TR:$args /F /RL:HIGHEST /SD:01/01/2050 /SC:ONCE /ST:00:00 ; schtasks.exe /run /TN:DTAConfig } -ArgumentList $exePath
+		Write-Verbose ("Error : {0} " -f ($err | out-string)) -Verbose
+		Write-Verbose ("Output : {0} " -f ($out | out-string)) -Verbose
+	}
+    Catch
+    {
+        Write-Verbose -Message ("Unable to start Agent process, will be rebooting the machine to complete the configuration") -Verbose
```
Comment (kmkumaran) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: I don't completely understand this. Would a catch always mean reboot of machine?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 6]
```diff
@@ -606,13 +606,52 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
...
+   
+    if( $ports -ne $null -and  $transports -ne $null -and $ports.Count -gt 0)
+    {
+        $port = $ports[0].InnerText
+        $transport = $transports[0].InnerText
+    }
+    else
+    {
+        Write-Verbose -Message("Unable to fetch WinRM config details. Using default port for configuration") -Verbose
+        $port = 5985
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: is this hardcoding intentional ?
scenario will break if anyone is using a non default port

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 (hunk 4 above)
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: +1

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 7]
```diff
@@ -606,13 +606,52 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
...
+        Write-Verbose -Message ("Unable to start Agent process, will be rebooting the machine to complete the configuration") -Verbose
+    }
+   
+}
 
-    Write-Verbose ("Error : {0} " -f ($err | out-string)) -Verbose
-    Write-Verbose ("Output : {0} " -f ($out | out-string)) -Verbose
+function CreateNewSession( [System.Management.Automation.PSCredential] $MachineCredentials)
+{
+    $winrmconfigDetails = Winrm e  winrm/config/listener -format:pretty |Out-String
```
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: my only concern here is, will the XML change with OS or in future?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 8]
```diff
@@ -606,13 +606,52 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
     }
     $exePath = Join-Path -Path $vsRoot -ChildPath $ExeName
     $exePath = "'" + $exePath + "'"
+	Try
```
Comment (ajay-arora) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: why not include the above lines of this method too ?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 (hunk 5 above)
Comment (acesiddhu) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: will this condition hit now?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 9]
```diff
@@ -727,6 +753,6 @@ $ret = CanSkipTestAgentConfiguration -TfsCollection $tfsCollectionUrl -AsService
 
 if ($ret -eq $false)
 {
-    $returnCode = ConfigureTestAgent -TfsCollection $tfsCollectionUrl -AsServiceOrProcess $asServiceOrProcess -EnvironmentUrl $environmentUrl -MachineName $machineName -MachineUserCredential $machineCredential -DisableScreenSaver $disableScreenSaver -EnableAutoLogon $enableAutoLogon -PersonalAccessToken $PersonalAccessToken -Capabilities $capabilities -AgentUserCredential $agentCredential
+    $returnCode = ConfigureTestAgent -TfsCollection $tfsCollectionUrl -AsServiceOrProcess $asServiceOrProcess -EnvironmentUrl $environmentUrl -MachineName $machineName -MachineUserCredential $machineCredential -DisableScreenSaver $disableScreenSaver -EnableAutoLogon $enableAutoLogon -PersonalAccessToken $PersonalAccessToken -Capabilities $capabilities -AgentUserCredential $agentCredential -WinrmPort $winrmport -WinrmTransport $WinrmTransport
```
Comment (kmkumaran) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: Where is this WinRM port coming for this script, specifically for this line?

File: Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1 [hunk 10]
```diff
@@ -606,13 +627,56 @@ function InvokeDTAExecHostExe([string] $Version, [System.Management.Automation.P
...
+	{  
+		$port = $ports[0].InnerText  
+		$transport = $transports[0].InnerText  
+	}  
+	else  
+	{  
+		Write-Verbose -Message("Unable to fetch WinRM config details. Using default port for configuration") -Verbose  
+		$port = 5985  
+		$transport = HTTP            
+	}
```
Comment (pavanadepu1) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: This might break if http is not configured at all on the machine.

Comment (pavanadepu1) [inline Tasks/DeployVisualStudioTestAgent/TestAgentConfiguration.ps1]: Also, validate the changes on different OS matrix. Especially on Win2K8 & Win7.

**Comment by aditya-MSFT** [comment]
Abondoning this pull request. Will be sending another one against master

**Pull request closed without merging**